File: src/threatforest/modules/utils/config_manager.py
"""Configuration management utilities"""

import shutil
from pathlib import Path
from typing import Any, Dict

import yaml
from questionary import confirm, select, text
from rich.console import Console
from rich.panel import Panel
from rich.table import Table

from threatforest.config import ROOT_DIR


class ConfigManager:
    """Manages ThreatForest configuration"""

    def __init__(self):
        self.console = Console()
        self.user_config_dir = ROOT_DIR / ".threatforest"
        self.user_config_file = self.user_config_dir / "config.yaml"
        self.bundled_config = Path(__file__).parent.parent.parent / "config.yaml"

    def init_user_config(self, force: bool = False) -> bool:
        """Initialize user config from bundled default"""
        if self.user_config_file.exists() and not force:
            self.console.print(f"[yellow]Config already exists:[/yellow] {self.user_config_file}")
            if not confirm("Overwrite existing config?", default=False).ask():
                return False

        # Create directory if needed
        self.user_config_dir.mkdir(parents=True, exist_ok=True)

        # Copy bundled config
        shutil.copy(self.bundled_config, self.user_config_file)

        self.console.print(
            f"\n[green]✓[/green] Created config: [cyan]{self.user_config_file}[/cyan]"
        )
        self.console.print("\n[dim]Edit this file to customize your ThreatForest settings.[/dim]\n")
        return True

    def show_config(self):
        """Display current configuration"""
        from threatforest.config import config

        # Determine which config is being used
        if (ROOT_DIR / ".threatforest" / "config.yaml").exists():
            config_source = f".threatforest/config.yaml (project config)"
        else:
            config_source = "Bundled default"

        # Detect active provider
        active_provider = "Not configured"
        model_id = "None"

        if config.bedrock and config.bedrock.get("model_id"):
            active_provider = "AWS Bedrock"
            model_id = config.bedrock.get("model_id")
        elif config.anthropic and config.anthropic.get("model_id"):
            active_provider = "Anthropic"
            model_id = config.anthropic.get("model_id")
        elif config.openai and config.openai.get("model_id"):
            active_provider = "OpenAI"
            model_id = config.openai.get("model_id")
        elif config.gemini and config.gemini.get("model_id"):
            active_provider = "Google Gemini"
            model_id = config.gemini.get("model_id")
        elif config.ollama and config.ollama.get("model_id"):
            active_provider = "Ollama"
            model_id = config.ollama.get("model_id")

        # Create table
        table = Table(title="ThreatForest Configuration", show_header=True)
        table.add_column("Setting", style="cyan", width=30)
        table.add_column("Value", style="green")

        # Add rows (no secrets like AWS profile)
        table.add_row("Config Source", config_source)
        table.add_row("Model Provider", active_provider)
        table.add_row("Model ID", model_id)
        table.add_row("Embeddings Model", config.embeddings_model)
        table.add_row("TTP Threshold", str(config.ttc_threshold))

        self.console.print()
        self.console.print(table)
        self.console.print()

    def edit_interactive(self):
        """Interactive configuration editor"""
        if not self.user_config_file.exists():
            self.console.print("[yellow]No user config found. Initializing...[/yellow]")
            self.init_user_config()

        # Load current config
        with open(self.user_config_file) as f:
            config_data = yaml.safe_load(f)

        self.console.print("\n[bold cyan]Interactive Configuration Editor[/bold cyan]\n")

        # Provider selection
        providers = [
            "AWS Bedrock",
            "Anthropic (Experimental)",
            "OpenAI (Experimental)",
            "Google Gemini (Experimental)",
            "Ollama (Experimental)",
            "Keep current",
        ]

        provider_choice = select("Select AI Provider:", choices=providers).ask()

        # Strip experimental flag from provider name
        provider_choice = provider_choice.replace(" (Experimental)", "")

        if provider_choice != "Keep current":
            self.console.print(f"\n[green]✓[/green] Selected: {provider_choice}")

            # Model selection based on provider
            if provider_choice == "AWS Bedrock":
                from threatforest.modules.utils.model_configs import BEDROCK_MODELS, DEFAULT_MODELS

                bedrock_models = BEDROCK_MODELS + ["Other (enter custom model ID)", "Keep current"]
                current_model = config_data.get("bedrock", {}).get(
                    "model_id", DEFAULT_MODELS["bedrock"]
                )
                model_choice = select(
                    f"Select Bedrock Model (current: {current_model}):", choices=bedrock_models
                ).ask()

                if model_choice != "Keep current":
                    # Check if user selected "Other" option
                    if "Other" in model_choice:
                        model_choice = text("Enter custom Bedrock model ID:", default="").ask()

                    if "bedrock" not in config_data:
                        config_data["bedrock"] = {}
                    config_data["bedrock"]["model_id"] = model_choice
                    self.console.print(f"[green]✓[/green] Model: {model_choice}")

            elif provider_choice == "Anthropic":
                from threatforest.modules.utils.model_configs import (
                    ANTHROPIC_MODELS,
                    DEFAULT_MODELS,
                )

                anthropic_models = ANTHROPIC_MODELS + ["Keep current"]
                current_model = config_data.get("anthropic", {}).get(
                    "model_id", DEFAULT_MODELS["anthropic"]
                )
                model_choice = select(
                    f"Select Anthropic Model (current: {current_model}):", choices=anthropic_models
                ).ask()

                if model_choice != "Keep current":
                    if "anthropic" not in config_data:
                        config_data["anthropic"] = {}
                    config_data["anthropic"]["model_id"] = model_choice
                    self.console.print(f"[green]✓[/green] Model: {model_choice}")

            elif provider_choice == "OpenAI":
                from threatforest.modules.utils.model_configs import DEFAULT_MODELS, OPENAI_MODELS

                openai_models = OPENAI_MODELS + ["Keep current"]
                current_model = config_data.get("openai", {}).get(
                    "model_id", DEFAULT_MODELS["openai"]
                )
                model_choice = select(
                    f"Select OpenAI Model (current: {current_model}):", choices=openai_models
                ).ask()

                if model_choice != "Keep current":
                    if "openai" not in config_data:
                        config_data["openai"] = {}
                    config_data["openai"]["model_id"] = model_choice
                    self.console.print(f"[green]✓[/green] Model: {model_choice}")

            elif provider_choice == "Google Gemini":
                from threatforest.modules.utils.model_configs import DEFAULT_MODELS, GEMINI_MODELS

                gemini_models = GEMINI_MODELS + ["Keep current"]
                current_model = config_data.get("gemini", {}).get(
                    "model_id", DEFAULT_MODELS["gemini"]
                )
                model_choice = select(
                    f"Select Gemini Model (current: {current_model}):", choices=gemini_models
                ).ask()

                if model_choice != "Keep current":
                    if "gemini" not in config_data:
                        config_data["gemini"] = {}
                    config_data["gemini"]["model_id"] = model_choice
                    self.console.print(f"[green]✓[/green] Model: {model_choice}")

            elif provider_choice == "Ollama (Local)":
                from threatforest.modules.utils.model_configs import DEFAULT_MODELS

                # For Ollama, allow custom model input
                current_model = config_data.get("ollama", {}).get(
                    "model_id", DEFAULT_MODELS["ollama"]
                )
                model_id = text(
                    f"Enter Ollama Model ID (current: {current_model}):", default=current_model
                ).ask()

                if "ollama" not in config_data:
                    config_data["ollama"] = {}
                config_data["ollama"]["model_id"] = model_id
                self.console.print(f"[green]✓[/green] Model: {model_id}")

        # Save changes
        with open(self.user_config_file, "w") as f:
            yaml.dump(config_data, f, default_flow_style=False, sort_keys=False)

        self.console.print(
            f"\n[green]✓[/green] Config saved: [cyan]{self.user_config_file}[/cyan]\n"
        )

    def set_value(self, key: str, value: str):
        """Set a specific configuration value"""
        if not self.user_config_file.exists():
            self.console.print("[yellow]No user config found. Initializing...[/yellow]")
            self.init_user_config()

        # Load config
        with open(self.user_config_file) as f:
            config_data = yaml.safe_load(f)

        # Parse dot-notation key
        keys = key.split(".")
        current = config_data

        # Navigate to parent
        for k in keys[:-1]:
            if k not in current:
                current[k] = {}
            current = current[k]

        # Set value
        current[keys[-1]] = value

        # Save
        with open(self.user_config_file, "w") as f:
            yaml.dump(config_data, f, default_flow_style=False, sort_keys=False)

        self.console.print(f"[green]✓[/green] Set [cyan]{key}[/cyan] = [yellow]{value}[/yellow]")

    def get_config_path(self) -> str:
        """Get path to active config file"""
        if self.user_config_file.exists():
            return str(self.user_config_file)
        return str(self.bundled_config)

File: ts/packages/cli/src/config-manager.ts
/**
 * Config CRUD — TS port of
 * `src/threatforest/modules/utils/config_manager.py`.
 *
 * Reads/writes `.threatforest/config.yaml`. The Python ConfigManager copies a
 * bundled `config.yaml`; the TS tree ships no such file, so init/migration build
 * a canonical default (the four SDK-native providers + the framework registry)
 * via `defaultConfigData()`. The resolution path matches the engine Config
 * (cwd-first), so the `threatforest run` wizard and `config show` agree.
 */
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';
import { parse as parseYaml, stringify as stringifyYaml } from 'yaml';
import { FRAMEWORKS } from '@threatforest/engine';
import { DEFAULT_MODELS } from './model-configs.js';
import * as display from './display.js';
import pc from 'picocolors';

/** The canonical default config (replaces copying a bundled config.yaml). */
export function defaultConfigData(): Record<string, unknown> {
  const frameworks: Record<string, unknown> = {};
  const descriptions: Record<string, string> = {
    attack: '835 techniques — cloud, network, endpoint',
    atlas: 'AI/ML adversarial threats',
    wiz: 'cloud threat landscape',
  };
  for (const [key, fw] of Object.entries(FRAMEWORKS)) {
    frameworks[key] = {
      name: fw.name,
      description: descriptions[key] ?? '',
      stix_bundle: fw.stix_bundle,
      source_name: fw.source_name,
      kill_chain_name: fw.kill_chain_name,
    };
  }
  return {
    bedrock: { model_id: DEFAULT_MODELS.bedrock },
    embeddings: { model: 'basel/ATTACK-BERT', ttc_threshold: 0.3 },
    frameworks,
  };
}

export class ConfigManager {
  readonly userConfigDir: string;
  readonly userConfigFile: string;

  constructor(rootDir: string = process.cwd()) {
    this.userConfigDir = join(rootDir, '.threatforest');
    this.userConfigFile = join(this.userConfigDir, 'config.yaml');
  }

  private read(): Record<string, unknown> {
    return (parseYaml(readFileSync(this.userConfigFile, 'utf8')) as Record<string, unknown>) ?? {};
  }

  private write(data: Record<string, unknown>): void {
    mkdirSync(this.userConfigDir, { recursive: true });
    writeFileSync(this.userConfigFile, stringifyYaml(data));
  }

  /** No-op migration hook (frameworks come from the canonical registry). */
  migrateConfig(): boolean {
    return false;
  }

  /** `config init` — write the default config (prompting before overwrite). */
  async initUserConfig(force = false): Promise<boolean> {
    if (existsSync(this.userConfigFile) && !force) {
      display.print(pc.yellow(`Config already exists: ${this.userConfigFile}`));
      const { confirm } = await import('@inquirer/prompts');
      const overwrite = await confirm({ message: 'Overwrite existing config?', default: false });
      if (!overwrite) return false;
    }
    this.write(defaultConfigData());
    display.blank();
    display.print(`${pc.green('✓')} Created config: ${pc.cyan(this.userConfigFile)}`);
    display.blank();
    display.print(pc.dim('Edit this file to customize your ThreatForest settings.'));
    display.blank();
    return true;
  }

  /** `config show` — render the active provider + key settings. */
  showConfig(): void {
    const data = existsSync(this.userConfigFile) ? this.read() : defaultConfigData();
    const configSource = existsSync(this.userConfigFile)
      ? '.threatforest/config.yaml (project config)'
      : 'Bundled default';

    const { provider, modelId } = detectProvider(data);
    const embeddings = (data.embeddings ?? {}) as Record<string, unknown>;
    const embeddingsModel = (embeddings.model as string) ?? 'basel/ATTACK-BERT';
    const ttc = (embeddings.ttc_threshold as number) ?? 0.3;

    const rows: [string, string][] = [
      ['Config Source', configSource],
      ['Model Provider', provider ?? 'Not configured'],
      ['Model ID', modelId ?? 'None'],
      ['Embeddings Model', embeddingsModel],
      ['TTP Threshold', String(ttc)],
    ];
    const keyWidth = Math.max(...rows.map(([k]) => k.length));
    display.blank();
    display.print(pc.bold(pc.cyan('ThreatForest Configuration')));
    for (const [k, v] of rows) display.print(`  ${pc.cyan(k.padEnd(keyWidth))}  ${pc.green(v)}`);
    display.blank();
  }

  /** `config edit` — interactive provider/model picker. */
  async editInteractive(): Promise<void> {
    if (!existsSync(this.userConfigFile)) {
      display.print(pc.yellow('No user config found. Initializing...'));
      await this.initUserConfig();
    }
    const data = this.read();
    display.blank();
    display.print(pc.bold(pc.cyan('Interactive Configuration Editor')));
    display.blank();

    const { select, input } = await import('@inquirer/prompts');
    const providerChoice = await select({
      message: 'Select AI Provider:',
      choices: [
        { name: 'AWS Bedrock', value: 'AWS Bedrock' },
        { name: 'Anthropic (Experimental)', value: 'Anthropic' },
        { name: 'OpenAI (Experimental)', value: 'OpenAI' },
        { name: 'Google Gemini (Experimental)', value: 'Google Gemini' },
        { name: 'Ollama (Experimental)', value: 'Ollama' },
        { name: 'Keep current', value: 'Keep current' },
      ],
    });

    if (providerChoice !== 'Keep current') {
      display.blank();
      display.print(`${pc.green('✓')} Selected: ${providerChoice}`);
      const { BEDROCK_MODELS, ANTHROPIC_MODELS, OPENAI_MODELS, GEMINI_MODELS } = await import(
        './model-configs.js'
      );

      const pickModel = async (
        blockKey: string,
        models: string[],
        allowOther: boolean,
      ): Promise<void> => {
        const block = (data[blockKey] ?? {}) as Record<string, unknown>;
        const current = (block.model_id as string) ?? DEFAULT_MODELS[blockKey];
        const choices = [
          ...models.map((m) => ({ name: m, value: m })),
          ...(allowOther ? [{ name: 'Other (enter custom model ID)', value: '__other__' }] : []),
          { name: 'Keep current', value: '__keep__' },
        ];
        let choice = await select({ message: `Select model (current: ${current}):`, choices });
        if (choice === '__keep__') return;
        if (choice === '__other__') {
          choice = await input({ message: 'Enter custom model ID:', default: '' });
        }
        data[blockKey] = { ...block, model_id: choice };
        display.print(`${pc.green('✓')} Model: ${choice}`);
      };

      if (providerChoice === 'AWS Bedrock') await pickModel('bedrock', BEDROCK_MODELS, true);
      else if (providerChoice === 'Anthropic') await pickModel('anthropic', ANTHROPIC_MODELS, false);
      else if (providerChoice === 'OpenAI') await pickModel('openai', OPENAI_MODELS, false);
      else if (providerChoice === 'Google Gemini') await pickModel('gemini', GEMINI_MODELS, false);
      else if (providerChoice === 'Ollama') {
        const block = (data.ollama ?? {}) as Record<string, unknown>;
        const current = (block.model_id as string) ?? DEFAULT_MODELS.ollama;
        const modelId = await input({ message: `Enter Ollama Model ID (current: ${current}):`, default: current });
        data.ollama = { ...block, model_id: modelId };
        display.print(`${pc.green('✓')} Model: ${modelId}`);
      }
    }

    this.write(data);
    display.blank();
    display.print(`${pc.green('✓')} Config saved: ${pc.cyan(this.userConfigFile)}`);
    display.blank();
  }

  /** `config set KEY VALUE` — dot-notation set. */
  async setValue(key: string, value: string): Promise<void> {
    if (!existsSync(this.userConfigFile)) {
      display.print(pc.yellow('No user config found. Initializing...'));
      await this.initUserConfig();
    }
    const data = this.read();
    const keys = key.split('.');
    let current: Record<string, unknown> = data;
    for (const k of keys.slice(0, -1)) {
      if (typeof current[k] !== 'object' || current[k] === null) current[k] = {};
      current = current[k] as Record<string, unknown>;
    }
    current[keys[keys.length - 1]!] = value;
    this.write(data);
    display.print(`${pc.green('✓')} Set ${pc.cyan(key)} = ${pc.yellow(value)}`);
  }

  /** `config path` — the active config file. */
  getConfigPath(): string {
    return this.userConfigFile;
  }
}

/** First-configured-block-wins provider detection (mirrors cli.py + config_manager.py). */
export function detectProvider(data: Record<string, unknown>): {
  provider: string | null;
  modelId: string | null;
} {
  const block = (k: string): Record<string, unknown> | null => {
    const v = data[k];
    return v && typeof v === 'object' ? (v as Record<string, unknown>) : null;
  };
  const order: [string, string, string][] = [
    ['bedrock', 'AWS Bedrock', 'model_id'],
    ['anthropic', 'Anthropic', 'model_id'],
    ['openai', 'OpenAI', 'model_id'],
    ['gemini', 'Google Gemini', 'model_id'],
    ['ollama', 'Ollama', 'model_id'],
    ['sagemaker', 'AWS SageMaker', 'endpoint_name'],
  ];
  for (const [key, name, idField] of order) {
    const b = block(key);
    if (b && b[idField]) return { provider: name, modelId: String(b[idField]) };
  }
  return { provider: null, modelId: null };
}
